contrib/babelfishpg_tsql/src/extendedproperty.c

```diff
@@ -0,0 +1,1139 @@
+	ScanKeyInit(&scanKey[3],
+				Anum_bbf_extended_properties_major_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(major_name));
+	ScanKeyInit(&scanKey[4],
+				Anum_bbf_extended_properties_minor_name,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(minor_name));
+	ScanKeyInit(&scanKey[5],
+				Anum_bbf_extended_properties_name,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(name));
+
+	scan = systable_beginscan(rel, get_bbf_extended_properties_idx_oid(), true,
+							  NULL, 6, scanKey);
+	if (HeapTupleIsValid(tuple = systable_getnext(scan)))
+	{
+		if (strcmp(procedure_name, "sp_addextendedproperty") == 0)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("Property cannot be added. Property '%s' already exists for '%s'.", name, var_object_name)));
+		}
+	}
+	else
+	{
+		if (strcmp(procedure_name, "sp_updateextendedproperty") == 0 ||
+			strcmp(procedure_name, "sp_dropextendedproperty") == 0)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("Property cannot be updated or deleted. Property '%s' does not exist for '%s'.", name, var_object_name)));
+		}
+	}
+
+	if (strcmp(procedure_name, "sp_addextendedproperty") == 0)
+	{
+		Datum		values[BBF_EXTENDED_PROPERTIES_NUM_COLS];
+		bool		nulls[BBF_EXTENDED_PROPERTIES_NUM_COLS];
+
+		MemSet(values, 0, sizeof(values));
+		MemSet(nulls, false, sizeof(nulls));
+
+		values[0] = Int16GetDatum(db_id);
+		values[1] = CStringGetDatum(schema_name);
+		values[2] = CStringGetDatum(major_name);
+		values[3] = CStringGetDatum(minor_name);
+		values[4] = CStringGetTextDatum(type);
+		values[5] = CStringGetTextDatum(name);
+		values[6] = CStringGetTextDatum(orig_name);
+		if (value)
+			values[7] = CStringGetDatum(value);
+		else
+			nulls[7] = true;
+
+		tuple = heap_form_tuple(RelationGetDescr(rel), values, nulls);
+		CatalogTupleInsert(rel, tuple);
+		heap_freetuple(tuple);
+	}
+	else if (strcmp(procedure_name, "sp_updateextendedproperty") == 0)
+	{
+		Datum		values[BBF_EXTENDED_PROPERTIES_NUM_COLS];
+		bool		nulls[BBF_EXTENDED_PROPERTIES_NUM_COLS];
+		bool		replaces[BBF_EXTENDED_PROPERTIES_NUM_COLS];
+		HeapTuple	new_tuple;
+
+		MemSet(values, 0, sizeof(values));
+		MemSet(nulls, false, sizeof(nulls));
+		MemSet(replaces, false, sizeof(replaces));
+
+		if (value)
+			values[Anum_bbf_extended_properties_value - 1] = CStringGetDatum(value);
+		else
+			nulls[Anum_bbf_extended_properties_value - 1] = true;
+		replaces[Anum_bbf_extended_properties_value - 1] = true;
+
+		new_tuple = heap_modify_tuple(tuple, RelationGetDescr(rel),
+									  values, nulls, replaces);
+		CatalogTupleUpdate(rel, &new_tuple->t_self, new_tuple);
+		heap_freetuple(new_tuple);
+	}
+	else if (strcmp(procedure_name, "sp_dropextendedproperty") == 0)
+	{
+		CatalogTupleDelete(rel, &tuple->t_self);
+	}
+
+	systable_endscan(scan);
+	table_close(rel, RowExclusiveLock);
+
+	CommandCounterIncrement();
+
+	if (type)
```

Comment:
Are they copied within this function? 
Usually we rely on the memory context, what is the consideration here?

Reply:
Refer to commonly realization in procedure.c. I think the code block has been in a memory context called "SPI exec".
If you really need memory context, please tell me directly.